File: integer-valued-list.go
package main

import "fmt"

func main() {
	// Load numbers
	numberLength := 0
	fmt.Scanf("%d", &numberLength)
	numbers := make([]int, numberLength)
	for i := 0; i < numberLength; i++ {
		fmt.Scanf("%d", &numbers[i])
	}

	// Calculate the total product
	var totalProduct int64 = 1
	countOfZero := 0
	for _, number := range numbers {
		if number != 0 {
			totalProduct *= int64(number)
		} else {
			if countOfZero++; countOfZero == 2 {
				break
			}
		}
	}

	// Print out results
	for _, number := range numbers {
		var output int64
		switch countOfZero {
		case 0:
			output = totalProduct / (int64(number))
		case 1:
			if number == 0 {
				output = totalProduct
			} else {
				output = 0
			}
		case 2:
			output = 0
		}

		fmt.Printf("%d\n", output)
	}
}

File: integer-valued-list.rb

number_length = gets.to_i
numbers = []
number_length.times.each{ numbers << gets.to_i }

count_of_zero = 0
total_product = 1
numbers.each do |number|
  total_product *= number if number != 0
  count_of_zero += 1 if number == 0
  break if count_of_zero > 1
end

numbers.each do |number|
  if count_of_zero > 1
    puts 0
  elsif count_of_zero == 1
    if number == 0
      puts total_product
    else
      puts 0
    end
  else
    puts total_product/number
  end
end
